### PR TITLE
Add a RAII type to manage the clipboard lifetime

### PR DESCRIPTION
**Pull request by 0xC0000054**
This removes the need for callers to use the `OpenClipboard` and `CloseClipboard` methods directly.

### CODE REVIEW

File: include/wil/resource.h [hunk 1]
```diff
@@ -4233,6 +4233,40 @@ namespace wil
...
+    struct unique_clipboard
+    {
+        unique_clipboard() = delete;
+
+        explicit unique_clipboard(_In_opt_ HWND hwndNewOwner) WI_NOEXCEPT
+            : openedClipboard(OpenClipboard(hwndNewOwner))
+        {
+        }
+		
+		// Cannot change memory location.
```
Comment (justanotheranonymoususer) [inline include/wil/resource.h:4246]: Lurking tabs

Comment (0xC0000054) [reply include/wil/resource.h]: Fixed

File: include/wil/resource.h [hunk 2]
```diff
@@ -4233,6 +4233,40 @@ namespace wil
 #if !defined(NOWINABLE)
     typedef unique_any<HWINEVENTHOOK, decltype(&::UnhookWinEvent), ::UnhookWinEvent> unique_hwineventhook;
 #endif
+#if !defined(NOCLIPBOARD)
```
Comment (jonwis) [inline include/wil/resource.h:4236]: This seems like a good fit for the "call method on close" helper, `wil::unique_call` plus a method, like:

```c++
using unique_clipboard = unique_call<decltype(::CloseClipboard()), &::CloseClipboard>;

unique_clipboard open_clipboard(HWND hwnd) {
    return { OpenClipboard(hwnd) != FALSE };
}
```

**Review by jonwis — CHANGES_REQUESTED** [review]

File: include/wil/resource.h (hunk 2 above)
Comment (0xC0000054) [reply include/wil/resource.h:4236]: I updated the code to use the `unique_call` type.

File: include/wil/resource.h [hunk 3]
```diff
@@ -4233,6 +4233,14 @@ namespace wil
 #if !defined(NOWINABLE)
     typedef unique_any<HWINEVENTHOOK, decltype(&::UnhookWinEvent), ::UnhookWinEvent> unique_hwineventhook;
 #endif
+#if !defined(NOCLIPBOARD)
+    using unique_clipboard = unique_call<decltype(::CloseClipboard()), &::CloseClipboard>;
+
+    inline unique_clipboard open_clipboard(HWND hwnd)
+	{
```
Comment (justanotheranonymoususer) [inline include/wil/resource.h:4240]: Tab among spaces

Comment (0xC0000054) [reply include/wil/resource.h]: Fixed.

**Review by jonwis — APPROVED** [review]

**Review by dunhor — APPROVED** [review]

File: include/wil/resource.h [hunk 4]
```diff
@@ -4233,6 +4233,14 @@ namespace wil
 #if !defined(NOWINABLE)
     typedef unique_any<HWINEVENTHOOK, decltype(&::UnhookWinEvent), ::UnhookWinEvent> unique_hwineventhook;
 #endif
+#if !defined(NOCLIPBOARD)
+    using unique_clipboard = unique_call<decltype(::CloseClipboard()), &::CloseClipboard>;
```
Comment (dunhor) [inline include/wil/resource.h:4237]: This should be `decltype(::CloseClipboard)` (not the result of the call)

Comment (dunhor) [reply include/wil/resource.h:4237]: Seems all of the existing `unique_call` aliases are named more of the form `unique_close_clipboard_call`. Probably good to maintain consistency.

**Review by raghavthind2005 — APPROVED** [review]

**Pull request merged**